File: python/scripts/create_plugin.py
#!/usr/bin/env python3

import argparse
from pathlib import Path

def create_project_toml(plugin_dir: Path, plugin_name: str, is_evm: bool, is_solana: bool = False) -> None:
    """Create the pyproject.toml file for the plugin."""
    # Base dependencies
    dependencies = '''python = "^3.10"
goat-sdk = "^0.1.0"'''
    
    # Add chain-specific dependencies
    if is_evm:
        dependencies += '\ngoat-sdk-wallet-evm = "^0.1.0"'
    elif is_solana:
        dependencies += '\ngoat-sdk-wallet-solana = "^0.1.0"'
    
    # Dev dependencies
    dev_dependencies = '''ruff = "^0.8.6"
goat-sdk = { path = "../../goat-sdk", develop = true }'''
    
    # Add chain-specific dev dependencies
    if is_evm:
        dev_dependencies += '\ngoat-sdk-wallet-evm = { path = "../../wallets/evm", develop = true }'
    elif is_solana:
        dev_dependencies += '\ngoat-sdk-wallet-solana = { path = "../../wallets/solana", develop = true }'
    
    toml_content = f'''[tool.poetry]
name = "goat-sdk-plugin-{plugin_name}"
version = "0.1.0"
description = "Goat plugin for {plugin_name}"
authors = ["Your Name <your_email@example.com>"]
readme = "README.md"
keywords = ["goat", "sdk", "agents", "ai", "{plugin_name}"]
homepage = "https://ohmygoat.dev/"
repository = "https://github.com/goat-sdk/goat"
packages = [
    {{ include = "goat_plugins/{plugin_name}" }},
]

[tool.poetry.dependencies]
{dependencies}

[tool.poetry.group.test.dependencies]
pytest = "^8.3.4"
pytest-asyncio = "^0.25.0"

[tool.poetry.urls]
"Bug Tracker" = "https://github.com/goat-sdk/goat/issues"

[tool.pytest.ini_options]
addopts = [
  "--import-mode=importlib",
]
pythonpath = "src"
asyncio_default_fixture_loop_scope = "function"

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"

[tool.poetry.group.dev.dependencies]
{dev_dependencies}

[tool.ruff]
line-length = 120
target-version = "py312"
'''
    
    with open(plugin_dir / "pyproject.toml", "w") as f:
        f.write(toml_content)

def create_parameters_file(goat_plugins_dir: Path, plugin_name: str) -> None:
    """Create the parameters.py file with example parameters in the goat_plugins directory."""
    parameters_content = '''from pydantic import BaseModel, Field
from typing import Optional


class ExampleQueryParameters(BaseModel):
    query: str = Field(
        description="An example query parameter (e.g., 'search term', 'identifier')"
    )
    limit: Optional[int] = Field(
        None,
        description="Maximum number of results to return. Defaults to no limit.",
    )
    include_metadata: bool = Field(
        default=False,
        description="Include additional metadata in the response"
    )


class ExampleActionParameters(BaseModel):
    target_id: str = Field(
        description="The ID of the target to perform action on"
    )
    action_type: str = Field(
        description="The type of action to perform (e.g., 'create', 'update', 'delete')"
    )
    parameters: Optional[dict] = Field(
        None,
        description="Optional parameters for the action"
    )
'''
    
    with open(goat_plugins_dir / "parameters.py", "w") as f:
        f.write(parameters_content)

def convert_to_python_identifier(name: str, for_class: bool = False) -> str:
    """Convert a kebab-case name to a valid Python identifier.
    
    Args:
        name: The name to convert
        for_class: If True, convert to PascalCase for class names,
                  otherwise convert to snake_case for function/variable names
    """
    # First convert to snake_case
    snake_case = name.replace("-", "_")
    
    if for_class:
        # Convert to PascalCase for class names
        return "".join(word.title() for word in snake_case.split("_"))
    
    return snake_case

def create_service_file(goat_plugins_dir: Path, plugin_name: str, is_evm: bool, is_solana: bool = False) -> None:
    """Create the service.py file with an empty tool."""
    # Convert plugin name to valid Python identifier for class name
    class_name = f"{convert_to_python_identifier(plugin_name, for_class=True)}Service"
    
    # Start with common imports
    service_content = '''from goat.decorators.tool import Tool
from .parameters import ExampleQueryParameters, ExampleActionParameters
'''

    # Add chain-specific imports if needed
    if is_evm:
        service_content += '''from goat_wallets.evm import EVMWalletClient

'''
    elif is_solana:
        service_content += '''from goat_wallets.solana import SolanaWalletClient

'''
    service_content += f'''
class {class_name}:
    def __init__(self, api_key: str):
        self.api_key = api_key

    @Tool({{
        "description": "Example query tool that demonstrates parameter usage",
        "parameters_schema": ExampleQueryParameters
    }})
    async def example_query(self{", wallet_client: EVMWalletClient" if is_evm else ", wallet_client: SolanaWalletClient" if is_solana else ""}, parameters: dict):
        """An example query method that shows how to use parameters."""
        try:
            # Example implementation
            query = parameters["query"]
            limit = parameters.get("limit")
            include_metadata = parameters.get("include_metadata", False)
            
            # Placeholder for actual implementation
            return {{"status": "success", "query": query, "limit": limit, "metadata_included": include_metadata}}
        except Exception as error:
            raise Exception(f"Failed to execute query: {{error}}")

    @Tool({{
        "description": "Example action tool that demonstrates parameter usage",
        "parameters_schema": ExampleActionParameters
    }})
    async def example_action(self{", wallet_client: EVMWalletClient" if is_evm else ", wallet_client: SolanaWalletClient" if is_solana else ""}, parameters: dict):
        """An example action method that shows how to use parameters."""
        try:
            # Example implementation
            target_id = parameters["target_id"]
            action_type = parameters["action_type"]
            action_params = parameters.get("parameters", {{}})
            
            # Placeholder for actual implementation
            return {{"status": "success", "action": action_type, "target": target_id, "params": action_params}}
        except Exception as error:
            raise Exception(f"Failed to execute action: {{error}}")
'''

    with open(goat_plugins_dir / "service.py", "w") as f:
        f.write(service_content)

def create_init_file(goat_plugins_dir: Path, plugin_name: str, is_evm: bool, is_solana: bool = False) -> None:
    """Create the __init__.py file with plugin class."""
    # Convert plugin name to valid Python identifiers
    python_name = convert_to_python_identifier(plugin_name)
    class_base = convert_to_python_identifier(plugin_name, for_class=True)
    class_name = f"{class_base}Service"
    plugin_class = f"{class_base}Plugin"
    options_class = f"{class_base}PluginOptions"
    
    init_content = '''from dataclasses import dataclass
from goat.classes.plugin_base import PluginBase
from .service import {class_name}


@dataclass
class {options_class}:
    """Options for the {plugin_class}."""
    api_key: str  # API key for external service integration


class {plugin_class}(PluginBase):
    def __init__(self, options: {options_class}):
        super().__init__("{plugin_name}", [{class_name}(options.api_key)])

    def supports_chain(self, chain) -> bool:
        return {supports_chain}


def {python_name}(options: {options_class}) -> {plugin_class}:
    return {plugin_class}(options)
'''.format(
        class_name=class_name,
        plugin_class=plugin_class,
        options_class=options_class,
        plugin_name=plugin_name,
        python_name=python_name,
        supports_chain="chain['type'] == 'evm'" if is_evm else "chain['type'] == 'solana'" if is_solana else "True"
    )

    with open(goat_plugins_dir / "__init__.py", "w") as f:
        f.write(init_content)

def main():
    """Main function to create a new GOAT plugin.
    
    This script generates a new plugin package for the GOAT SDK with a standardized structure.
    The plugin name should use hyphens for spaces (e.g., 'my-plugin') and be lowercase.
    Class names will automatically convert hyphens to underscores for valid Python syntax.
    
    Examples:
        # Create an EVM-compatible plugin
        python create_plugin.py my-token --evm
        
        # Create a chain-agnostic plugin
        python create_plugin.py my-service
        
    The script will create:
    1. pyproject.toml with proper dependencies
    2. goat_plugins directory with:
       - parameters.py: Example Pydantic models
       - service.py: Service class with Tool decorators
       - __init__.py: Plugin class and initialization
    3. README.md with usage instructions
    """
    parser = argparse.ArgumentParser(description="Create a new GOAT plugin.")
    parser.add_argument("name", help="Plugin name, e.g. 'myplugin'")
    parser.add_argument("--evm", action="store_true", help="Indicate if plugin is for EVM")
    parser.add_argument("--solana", action="store_true", help="Indicate if plugin is for Solana")
    args = parser.parse_args()
    
    # Ensure only one chain type is specified
    if args.evm and args.solana:
        parser.error("Please specify only one chain type (--evm or --solana)")
    
    plugin_name = args.name.lower()  # Normalize to lowercase
    is_evm = args.evm
    is_solana = args.solana
    
    # Convert plugin name to valid Python identifiers (used throughout the script)
    python_name = convert_to_python_identifier(plugin_name)
    class_base = convert_to_python_identifier(plugin_name, for_class=True)
    
    # Create base plugin directory (relative to python root)
    plugin_dir = Path(__file__).parent.parent / "src" / "plugins" / plugin_name
    plugin_dir.mkdir(parents=True, exist_ok=True)
    
    # Create all required files
    create_project_toml(plugin_dir, plugin_name, is_evm, is_solana)
    
    # Create goat_plugins directory and its contents
    goat_plugins_dir = plugin_dir / "goat_plugins" / plugin_name
    goat_plugins_dir.mkdir(parents=True, exist_ok=True)
    
    # Create all plugin files
    create_parameters_file(goat_plugins_dir, plugin_name)
    create_service_file(goat_plugins_dir, plugin_name, is_evm, is_solana)
    create_init_file(goat_plugins_dir, plugin_name, is_evm, is_solana)
    
    # Create README.md
    # Convert plugin name for Python usage
    python_name = convert_to_python_identifier(plugin_name)
    
    # Determine chain support text
    chain_support = 'EVM chain support' if is_evm else 'Solana chain support' if is_solana else 'Chain-agnostic support'
    
    # Determine additional dependencies
    additional_deps = ''
    if is_evm:
        additional_deps = '# Install required wallet dependency\npoetry add goat-sdk-wallet-evm'
    elif is_solana:
        additional_deps = '# Install required wallet dependency\npoetry add goat-sdk-wallet-solana'
    
    readme_content = f"""# {plugin_name} Plugin for GOAT SDK

A plugin for the GOAT SDK that provides {plugin_name} functionality.

## Installation

```bash
# Install the plugin
poetry add goat-sdk-plugin-{plugin_name}

{additional_deps if additional_deps else ''}
```

## Usage

```python
from goat_plugins.{plugin_name} import {python_name}, {class_base}PluginOptions

# Initialize the plugin
options = {class_base}PluginOptions(
    api_key="your-api-key"
)
plugin = {python_name}(options)
```

## Features

- Example query functionality
- Example action functionality
- {chain_support}

## License

This project is licensed under the terms of the MIT license.
"""
    with open(plugin_dir / "README.md", "w") as f:
        f.write(readme_content)
    
    print(f"Plugin '{plugin_name}' created successfully in {plugin_dir}")

if __name__ == "__main__":
    main()

File: typescript/scripts/createPlugin.ts
#!/usr/bin/env node
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { Command } from "commander";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

type PluginType = "evm" | "solana" | "any";

const program = new Command();

interface PluginOptions {
    name: string;
    type: "evm" | "solana" | "any";
}

// Validate plugin name format
function validatePluginName(name: string): boolean {
    return /^[a-z0-9-]+$/.test(name);
}

// Get wallet client import based on plugin type
function getWalletClientImport(type: PluginOptions["type"]): string {
    switch (type) {
        case "evm":
            return 'import { EVMWalletClient } from "@goat-sdk/wallet-evm";';
        case "solana":
            return 'import { SolanaWalletClient } from "@goat-sdk/wallet-solana";';
        case "any":
            return 'import { WalletClientBase } from "@goat-sdk/core";';
    }
}

// Get wallet client type based on plugin type
function getWalletClientType(type: PluginOptions["type"]): string {
    switch (type) {
        case "evm":
            return "EVMWalletClient";
        case "solana":
            return "SolanaWalletClient";
        case "any":
            return "WalletClientBase";
    }
}

// Get dependencies based on plugin type
function getDependencies(type: PluginOptions["type"]): Record<string, string> {
    const deps: Record<string, string> = {
        "@goat-sdk/core": "workspace:*",
        zod: "catalog:",
    };

    switch (type) {
        case "evm":
            deps["@goat-sdk/wallet-evm"] = "workspace:*";
            break;
        case "solana":
            deps["@goat-sdk/wallet-solana"] = "workspace:*";
            break;
    }

    return deps;
}

// Create package.json content
function createPackageJson(options: PluginOptions): string {
    const { name, type } = options;
    const dependencies = getDependencies(type);

    const packageJson = {
        name: `@goat-sdk/plugin-${name}`,
        version: "0.1.0",
        files: ["dist/**/*", "README.md", "package.json"],
        scripts: {
            build: "tsup",
            clean: "rm -rf dist",
            test: "vitest run --passWithNoTests",
        },
        main: "./dist/index.js",
        module: "./dist/index.mjs",
        types: "./dist/index.d.ts",
        sideEffects: false,
        homepage: "https://ohmygoat.dev",
        repository: {
            type: "git",
            url: "git+https://github.com/goat-sdk/goat.git",
        },
        license: "MIT",
        bugs: {
            url: "https://github.com/goat-sdk/goat/issues",
        },
        keywords: ["ai", "agents", "web3"],
        dependencies,
        peerDependencies: {
            "@goat-sdk/core": "workspace:*",
        },
    };
    return JSON.stringify(packageJson, null, 4);
}

// Create parameters.ts content
function createParametersContent(): string {
    return `import { createToolParameters } from "@goat-sdk/core";
import { z } from "zod";

export class ExampleParameters extends createToolParameters(
    z.object({
        exampleField: z.string().describe("An example field"),
    }),
) {}
`;
}

// Convert kebab-case to PascalCase
function kebabToPascalCase(str: string): string {
    return str
        .split("-")
        .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
        .join("");
}

// Create service.ts content
function createServiceContent(options: PluginOptions): string {
    const { name, type } = options;
    const walletClientImport = getWalletClientImport(type);
    const walletClientType = getWalletClientType(type);
    const className = kebabToPascalCase(name);
    const functionName = name.replace(/-/g, "_");

    return `import { Tool } from "@goat-sdk/core";
${walletClientImport}
import { ExampleParameters } from "./parameters";

export class ${className}Service {
    @Tool({
        name: "${functionName}_example",
        description: "An example method in ${className}Service",
    })
    async doSomething(walletClient: ${walletClientType}, parameters: ExampleParameters) {
        // Implementation goes here
        return "Hello from ${className}Service!";
    }
}
`;
}

// Create plugin.ts content
function createPluginContent(options: PluginOptions): string {
    const { name } = options;
    const className = kebabToPascalCase(name);
    const functionName = name.replace(/-/g, "");

    return `import { PluginBase } from "@goat-sdk/core";
import { ${className}Service } from "./${name}.service";

export class ${className}Plugin extends PluginBase {
    constructor() {
        super("${name}", [new ${className}Service()]);
    }

    supportsChain = () => true;
}

export function ${functionName}() {
    return new ${className}Plugin();
}
`;
}

// Create index.ts content
function createIndexContent(options: PluginOptions): string {
    const { name } = options;
    return `export * from "./${name}.plugin";
export * from "./parameters";
`;
}

// Create README.md content
function createReadmeContent(options: PluginOptions): string {
    const { name } = options;
    const className = kebabToPascalCase(name);
    const functionName = name.replace(/-/g, "");

    return `<div align="center">
<a href="https://github.com/goat-sdk/goat">

<img src="https://github.com/user-attachments/assets/5fc7f121-259c-492c-8bca-f15fe7eb830c" alt="GOAT" width="100px" height="auto" style="object-fit: contain;">
</a>
</div>

# ${className} GOAT Plugin

Brief description of the ${className} plugin and what it does

## Installation
\`\`\`bash
npm install @goat-sdk/plugin-${name}
yarn add @goat-sdk/plugin-${name}
pnpm add @goat-sdk/plugin-${name}
\`\`\`

## Usage
\`\`\`typescript
import { ${functionName} } from '@goat-sdk/plugin-${name}';

const tools = await getOnChainTools({
    wallet: // ...
    plugins: [
       ${functionName}()
    ]
});
\`\`\`

## Tools
* Tool 1
* Tool 2
* Tool 3

<footer>
<br/>
<br/>
<div>
<a href="https://github.com/goat-sdk/goat">
  <img src="https://github.com/user-attachments/assets/59fa5ddc-9d47-4d41-a51a-64f6798f94bd" alt="GOAT" width="100%" height="auto" style="object-fit: contain; max-width: 800px;">
</a>
</div>
</footer>
`;
}

// Create all necessary files for the plugin
async function createPlugin(options: PluginOptions): Promise<void> {
    const { name } = options;
    const rootDir = process.cwd();
    const pluginDir = path.join(rootDir, "packages", "plugins", name);
    const srcDir = path.join(pluginDir, "src");

    // Create directories
    fs.mkdirSync(pluginDir, { recursive: true });
    fs.mkdirSync(srcDir, { recursive: true });

    // Copy configuration files from superfluid plugin
    const superfluidDir = path.join(rootDir, "packages", "plugins", "superfluid");
    const configFiles = ["tsconfig.json", "tsup.config.ts", "turbo.json"];

    for (const file of configFiles) {
        const sourcePath = path.join(superfluidDir, file);
        const targetPath = path.join(pluginDir, file);
        try {
            if (fs.existsSync(sourcePath)) {
                fs.copyFileSync(sourcePath, targetPath);
            } else {
                console.warn(`Warning: Could not find ${file} in superfluid plugin`);
            }
        } catch (error) {
            console.error(`Error copying ${file}:`, error);
            throw error;
        }
    }

    // Create all files first
    console.log("Creating plugin files...");
    // Create package.json
    const packageJsonPath = path.join(pluginDir, "package.json");
    fs.writeFileSync(packageJsonPath, createPackageJson(options));
    console.log("✓ Created package.json");

    // Create README.md
    const readmePath = path.join(pluginDir, "README.md");
    fs.writeFileSync(readmePath, createReadmeContent(options));
    console.log("✓ Created README.md");

    // Create source files
    const sourceFiles = {
        "parameters.ts": createParametersContent(),
        [`${name}.service.ts`]: createServiceContent(options),
        [`${name}.plugin.ts`]: createPluginContent(options),
        "index.ts": createIndexContent(options),
    };

    for (const [file, content] of Object.entries(sourceFiles)) {
        const filePath = path.join(srcDir, file);
        fs.writeFileSync(filePath, content);
        console.log(`✓ Created ${file}`);
    }

    // Verify all required files exist
    const requiredFiles = [
        packageJsonPath,
        path.join(pluginDir, "tsconfig.json"),
        path.join(pluginDir, "tsup.config.ts"),
        path.join(pluginDir, "turbo.json"),
        readmePath,
        ...Object.keys(sourceFiles).map((file) => path.join(srcDir, file)),
    ];

    console.log("\nVerifying required files...");
    for (const file of requiredFiles) {
        if (!fs.existsSync(file)) {
            throw new Error(`Required file missing: ${file}`);
        }
        console.log(`✓ Verified ${path.basename(file)}`);
    }

    // Install dependencies and build
    try {
        console.log("\nInstalling dependencies...");
        const { execSync } = await import("node:child_process");
        await execSync("pnpm install", { stdio: "inherit", cwd: rootDir });
        console.log("✓ Dependencies installed");
        console.log("\nBuilding plugin...");
        await execSync("pnpm build", { stdio: "inherit", cwd: pluginDir });
        console.log("✓ Plugin built successfully");
    } catch (error) {
        console.error("Error during install/build:", error);
        throw error;
    }

    console.log(`\nPlugin ${name} created successfully!`);
}

// Set up command line interface
program
    .name("createPlugin")
    .description("Create a new GOAT SDK plugin")
    .requiredOption("-n, --name <name>", "plugin name (lowercase, hyphen-separated)")
    .option("-t, --type <type>", "plugin type (evm, solana, or any)", "any")
    .action(async (options: { name: string; type: string }) => {
        const { name, type } = options;

        // Validate inputs
        if (!validatePluginName(name)) {
            console.error("Error: Plugin name must be lowercase and may only contain letters, numbers, and hyphens");
            process.exit(1);
        }

        const validTypes = ["evm", "solana", "any"] as const;
        const pluginType = type || "any";

        if (!validTypes.includes(pluginType as (typeof validTypes)[number])) {
            console.error("Error: Plugin type must be one of: evm, solana, any");
            process.exit(1);
        }

        try {
            await createPlugin({
                name,
                type: pluginType as PluginOptions["type"],
            });
        } catch (error) {
            console.error("Error creating plugin:", error);
            process.exit(1);
        }
    });

program.parse();
